src/hotspot/cpu/x86/c2_stubGenerator_x86_64_string.cpp

```diff
@@ -1672,6 +1629,86 @@ static void highly_optimized_short_cases(StrIntrinsicNode::ArgEncoding ae, Regis
   __ jmpb(L_out);
 }
 
+
+
+// Copy the small (< 32 byte) haystack to the stack.  Allows for vector reads without page fault
```

Comment:
Just to be pedantic, its `(<=32)` - this function also handles 32bytes case.

-  line 401: 
```
 __ cmpq(haystack_len, 0x20);
 __ ja(L_bigSwitchTop);
```
- though line 293 (`highly_optimized_short_cases`) only seems to route16-byte cases here: 
```__ cmpq(haystack_len_p, isU ? 8 : 16);```

Reply:
I am not sure what you are looking at, but line 293 reads:
```
__ cmpq(haystack_len_p, isU ? 16 : 32);
```

for me. IOW, it routes > 32 byte cases to `L_begin`. But the following cmp/ja also routes <= 32 byte cases there, when `needle_len > 6`.